modules/fx/arrow-fx/src/main/kotlin/arrow/fx/internal/CancelableQueue.kt

```diff
@@ -0,0 +1,366 @@
+        if (state.compareAndSet(current, State.Deficit(newReads, current.takes, current.shutdownHook))) later { unsafeCancelRead(id) }
+        else unsafeRead(onRead)
+      }
+      is State.Shutdown -> raiseError(QueueShutdown)
+    }
+
+  private tailrec fun unsafeCancelRead(id: Token): Unit =
+    when (val current = state.value) {
+      is State.Deficit -> {
+        val newMap = current.reads - id
+        val update = State.Deficit<F, A>(newMap, current.takes, current.shutdownHook)
+        if (state.compareAndSet(current, update)) Unit
+        else unsafeCancelRead(id)
+      }
+      else -> Unit
+    }
+
+  private fun callPutAndAllReaders(a: A, put: ((Either<Nothing, A>) -> Unit)?, reads: Map<Token, (Either<Nothing, A>) -> Unit>): Kind<F, Boolean> {
+    val value = Right(a)
+    return reads.values.callAll(value).flatMap {
+      if (put != null) later { put(value) }.fork(EmptyCoroutineContext).map { true }
+      else just(true)
+    }
+  }
+
+  // For streaming a value to a whole `reads` collection
+  private fun Iterable<(Either<Nothing, A>) -> Unit>.callAll(value: Either<Nothing, A>): Kind<F, Unit> =
+    fold(null as Kind<F, Fiber<F, Unit>>?) { acc, cb ->
+      val task = later { cb(value) }.fork(EmptyCoroutineContext)
+      acc?.flatMap { task } ?: task
+    }?.map(mapUnit) ?: unit()
+
+  override fun <A, B> Kind<F, A>.ap(ff: Kind<F, (A) -> B>): Kind<F, B> = CF.run {
+    this@ap.ap(ff)
+  }
+
+  override fun <A, B> Kind<F, A>.map(f: (A) -> B): Kind<F, B> = CF.run {
+    this@map.map(f)
+  }
+
+  internal sealed class SurplusStrategy<F, A> {
+    abstract fun unsafeTryHandleSurplus(state: AtomicRefW<State<F, A>>, a: A, surplus: State.Surplus<F, A>): Kind<F, Boolean>?
+
+    data class Bounded<F, A>(val capacity: Int, val AP: Applicative<F>) : SurplusStrategy<F, A>() {
+      override fun unsafeTryHandleSurplus(state: AtomicRefW<State<F, A>>, a: A, surplus: State.Surplus<F, A>): Kind<F, Boolean>? = when {
+        surplus.value.length() >= capacity -> AP.just(false)
+        state.compareAndSet(surplus, surplus.copy(surplus.value.enqueue(a))) -> AP.just(true)
+        else -> null
+      }
+    }
+
+    data class Sliding<F, A>(val capacity: Int, val AP: Applicative<F>) : SurplusStrategy<F, A>() {
+      override fun unsafeTryHandleSurplus(state: AtomicRefW<State<F, A>>, a: A, surplus: State.Surplus<F, A>): Kind<F, Boolean>? {
+        val nextQueue = if (surplus.value.length() < capacity) surplus.value.enqueue(a) else surplus.value.dequeue().b.enqueue(a)
+        return if (!state.compareAndSet(surplus, surplus.copy(value = nextQueue))) null
+        else AP.just(true)
+      }
+    }
+
+    data class Dropping<F, A>(val capacity: Int, val AP: Applicative<F>) : SurplusStrategy<F, A>() {
+      override fun unsafeTryHandleSurplus(state: AtomicRefW<State<F, A>>, a: A, surplus: State.Surplus<F, A>): Kind<F, Boolean>? {
+        val nextQueue = if (surplus.value.length() < capacity) surplus.value.enqueue(a) else surplus.value
```

Comment:
NIT: brackets on ifs

Reply:
Is this something we can configure in KtLint?
I'm used to using expressions for `if/else`, and probably in more places than I should...